consensus/replay_test.go

```diff
@@ -278,42 +281,194 @@ var (
+
+	//---------------------------------------------------------------------------
+	logger.Info("---------------------------- Testing changing the voting power of one validator")
+
+	updateValidatorPubKey1 := css[nVals].privValidator.GetPubKey()
+	updatePubKey1ABCI := types.TM2PB.PubKey(updateValidatorPubKey1)
+	updateValidatorTx1 := kvstore.MakeValSetChangeTx(updatePubKey1ABCI, 25)
+
+	//height 3‘s block contains newValidatorTx1. It will make effect in height 5, will appear in height 6's block's commit
+	waitForAndValidateBlock(t, nPeers, activeVals, blocksSubs, css, updateValidatorTx1)
+
+	//---------------------------------------------------------------------------
+	logger.Info("---------------------------- Testing adding two validators at once")
+
+	newValidatorPubKey2 := css[nVals+1].privValidator.GetPubKey()
+	newVal2ABCI := types.TM2PB.PubKey(newValidatorPubKey2)
+	newValidatorTx2 := kvstore.MakeValSetChangeTx(newVal2ABCI, testMinPower)
+
+	newValidatorPubKey3 := css[nVals+2].privValidator.GetPubKey()
+	newVal3ABCI := types.TM2PB.PubKey(newValidatorPubKey3)
+	newValidatorTx3 := kvstore.MakeValSetChangeTx(newVal3ABCI, testMinPower)
+
+	//height 4
+	waitForAndValidateBlock(t, nPeers, activeVals, blocksSubs, css, newValidatorTx2, newValidatorTx3)
+
+	//---------------------------------------------------------------------------
+	logger.Info("---------------------------- Testing removing two validators at once")
+
+	removeValidatorTx2 := kvstore.MakeValSetChangeTx(newVal2ABCI, 0)
+	removeValidatorTx3 := kvstore.MakeValSetChangeTx(newVal3ABCI, 0)
+
+	//height 5
+	waitForAndValidateBlock(t, nPeers, activeVals, blocksSubs, css, removeValidatorTx2, removeValidatorTx3)
+	activeVals[string(newValidatorPubKey1.Address())] = struct{}{}
+	//height 6
+	waitForBlockWithUpdatedValsAndValidateIt(t, nPeers, activeVals, blocksSubs, css)
+	stopConsensusNet(logger, reactors, eventBuses)
+
+	sim_chain = make([]*types.Block, 0)
+	sim_commits = make([]*types.Commit, 0)
+	for i := 1; i <= NUM_BLOCKS; i++ {
+		sim_chain = append(sim_chain, css[0].blockStore.LoadBlock(int64(i)))
+		sim_commits = append(sim_commits, css[0].blockStore.LoadBlockCommit(int64(i)))
+	}
+}
+
 // Sync from scratch
 func TestHandshakeReplayAll(t *testing.T) {
 	for i, m := range modes {
-		config := ResetConfig(fmt.Sprintf("%s_%v", t.Name(), i))
+		config := ResetConfig(fmt.Sprintf("%s_%v_s", t.Name(), i))
 		defer os.RemoveAll(config.RootDir)
-		testHandshakeReplay(t, config, 0, m)
+		testHandshakeReplay(t, config, 0, m, false)
+	}
+	for i, m := range modes {
+		config := ResetConfig(fmt.Sprintf("%s_%v_m", t.Name(), i))
+		defer os.RemoveAll(config.RootDir)
+		testHandshakeReplay(t, config, 0, m, true)
 	}
 }
 
 // Sync many, not from scratch
 func TestHandshakeReplaySome(t *testing.T) {
 	for i, m := range modes {
-		config := ResetConfig(fmt.Sprintf("%s_%v", t.Name(), i))
+		config := ResetConfig(fmt.Sprintf("%s_%v_s", t.Name(), i))
```

Comment:
This set of three lines is repeated so often we should make it its own function? then we can just call it twice for each iteration (once with false, once with true)